Enhanced parsing of input values (post or parameters)

jlannoy opened the pull request:
Hello.

I already tried to send this [request](https://github.com/ninjaframework/ninja/pull/448) before but your review, the modifications needed and the evolution of the main repository I have not been able to do a rebase. I tried several times but finally it took less pain to copy them on a fresh branch. 

If you remember it, the goal of this PR is to enhance the parsing of strings to simple types, from the parameters or from the request. In details, it allows :
- To avoid duplicated code or differences between ParamParsers and SwissKnife converters (their job are in fact the same)
- To allow the POST body engine to register validation errors on conversion failure
- To avoid registering enumeration at start
- To allow users to register their own ParamParser
- Even overriding default ones, which can be useful to specify a particular DateFormat
- To allow parsing of other objects inside the parsed POST object

Notes :
- Custom ParamParsers are instancied via an Injector, so that users can make usage of @Inject dependencies
- The convert method for Date from the SwissKnife class have been translated to a ParamParser, based on LocalDateTime (as the original method). I’m thinking about something more useful, with date format from the properties file based on current Locale ; but I didn’t want to make a bigger PR. The fact is that with this PR, each user will be able to write a parser with it’s own format.
- It works with the MultiBinder, and is documented as well
- I had to enhanced the Validation documentation page, to make the differences with the JSR303 validator clear (interpolated messages in bean violation // message keys in field validations)

Regards,

Jonathan

jlannoy committed:
Merge remote-tracking branch 'ninjaframework/develop' into feature_bodyparser_improvement2

jlannoy committed:
Body Parsing Improvement

Documentation for Validation enhanced. SwissKnife converters replaced
by usage of ParamParsers in BodyPostEngine. Support for enumeration
value (registration deprecated), custom types parsing and POST objects
parsing.

jlannoy commented:
Hello @raphaelbauer and @jjlauer ... Don't forget this change. It was quite hard to merge my modifications from the previous pull request, I'd rather not doing it again. Thanks!

raphaelbauer commented:
Hi @jlannoy thanks for bringing that PR back to our attention! @jjlauer I guess you had a discussion about the PR already - would you like to handle it or should I do it?

jjlauer commented:
@raphaelbauer I am compressed on time these days so it'd be great if you could handle it!

raphaelbauer commented:
Cool. Will do. @jlannoy I will handle the PR in the coming days :)

raphaelbauer commented on ninja-core/src/main/java/ninja/Bootstrap.java:
```diff
@@ -208,6 +209,7 @@ protected void configure() throws Exception {
             @Override
             protected void configure() {
                 bind(Ninja.class).to(ninjaClass).in(Singleton.class);
+                Multibinder.newSetBinder(binder(), ParamParser.class);
```
Not sure, but either we add the binder here or in NinjaBaseModule - but not in both... I guess NinjaBaseModule is the right place...

jlannoy replied:
I completely agree... I think I did it twice because I started this PR before the NinjaBaseModule class appeared ;-)

raphaelbauer commented on ninja-core/src/main/java/ninja/bodyparser/BodyParserEnginePost.java:
```diff
@@ -49,46 +65,67 @@
         }
 
-        Set<String> declaredFields = getAllDeclaredFieldsAsStringSet(classOfT);
+        for(String declaredField : getAllDeclaredFieldsAsStringSet(classOfT)) {
```
tiny, but there should be a space between for and ( like "for ("...

raphaelbauer commented on ninja-core/src/main/java/ninja/bodyparser/BodyParserEnginePost.java:
```diff
@@ -49,46 +65,67 @@
+                } else {
+                    
+                    // Check if we have a parameter key corresponding to inner attributes of this field
+                    for(String parameter : context.getParameters().keySet()) {
```
Again very tiny, but there should be a space between for and ( like "for ("...

raphaelbauer commented on ninja-core/src/main/java/ninja/params/ParamParsers.java:
```diff
@@ -462,46 +492,68 @@ public Character parseParameter(String field, String parameterValue, Validation
-                        }
+//                try {
+//                    return Enum.valueOf(targetType, parameterValue);
+//                }
```
I think you can remove the block that is commented out...

raphaelbauer commented:
Hey @jlannoy that PR is very well done. Puh. What a great work on so many parts of Ninja :) Congrats!

There are some minor things I annotated. Once these are resolved we'll merge the PR (and make a new Ninja release...)

jlannoy committed:
Fix after Raphael review

Two spaces added between for and ( ; commented block removed ;
duplicate multibinder instantiation fixed

jlannoy commented:
Thank you @raphaelbauer. I fixed the 4 comments, you can check the last commit ;-)

raphaelbauer commented:
Perfect - Thanks!
